op-batcher/batcher/sync_actions.go

```diff
@@ -0,0 +1,144 @@
+package batcher
+
+import (
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/queue"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type channelStatuser interface {
+	isFullySubmitted() bool
+	isTimedOut() bool
+	LatestL2() eth.BlockID
+	MaxInclusionBlock() uint64
+}
+
+type syncActions struct {
+	clearState      *eth.BlockID
+	blocksToPrune   int
+	channelsToPrune int
+	blocksToLoad    [2]uint64 // the range [start,end] that should be loaded into the local state.
+	// NOTE this range is inclusive on both ends, which is a change to previous behaviour.
+}
+
+func (s syncActions) String() string {
+	return fmt.Sprintf(
+		"SyncActions{blocksToPrune: %d, channelsToPrune: %d, clearState: %v, blocksToLoad: [%d, %d]}", s.blocksToPrune, s.channelsToPrune, s.clearState, s.blocksToLoad[0], s.blocksToLoad[1])
+}
+
+// computeSyncActions determines the actions that should be taken based on the inputs provided. The inputs are the current
+// state of the batcher (blocks and channels), the new sync status, and the previous current L1 block. The actions are returned
+// in a struct specifying the number of blocks to prune, the number of channels to prune, whether to wait for node sync, the block
+// range to load into the local state, and whether to clear the state entirely. Returns an boolean indicating if the sequencer is out of sync.
+func computeSyncActions[T channelStatuser](newSyncStatus eth.SyncStatus, prevCurrentL1 eth.L1BlockRef, blocks queue.Queue[*types.Block], channels []T, l log.Logger) (syncActions, bool) {
+
+	// PART 1: Initial checks on the sync status
+	if newSyncStatus.HeadL1 == (eth.L1BlockRef{}) {
+		l.Warn("empty sync status")
+		return syncActions{}, true
+	}
+
+	if newSyncStatus.CurrentL1.Number < prevCurrentL1.Number {
+		// This can happen when the sequencer restarts
+		l.Warn("sequencer currentL1 reversed")
+		return syncActions{}, true
+	}
+
+	// PART 2: checks involving only the oldest block in the state
+	oldestBlockInState, hasBlocks := blocks.Peek()
+	oldestUnsafeBlockNum := newSyncStatus.SafeL2.Number + 1
+	youngestUnsafeBlockNum := newSyncStatus.UnsafeL2.Number
+
+	if !hasBlocks {
+		s := syncActions{
+			blocksToLoad: [2]uint64{oldestUnsafeBlockNum, youngestUnsafeBlockNum},
+		}
+		l.Info("no blocks in state", "syncActions", s)
+		return s, false
+	}
+
+	oldestBlockInStateNum := oldestBlockInState.NumberU64()
+
+	if oldestUnsafeBlockNum < oldestBlockInStateNum {
+		s := syncActions{
+			clearState:   &newSyncStatus.SafeL2.L1Origin,
+			blocksToLoad: [2]uint64{oldestUnsafeBlockNum, youngestUnsafeBlockNum},
+		}
```

Comment:
you could also use `startAfresh` here.

Reply:
[c534f92](https://github.com/ethereum-optimism/optimism/pull/13060/commits/c534f92540a70a39cbda91d1d475ef1f0fe7df25)